### PR TITLE
GH-240 Add IsMainThread detection API

### PR DESCRIPTION
**Pull request by jamesmontemagno**
### Description of Change ###

Describe your changes here. 

### Bugs Fixed ###

- Related to issue #240

Provide links to issues here. Ensure that a GitHub issue was created for your feature or bug fix before sending PR.

### API Changes ###

Added: 
 
- bool- boolPlatform.IsMainThread { get; } 

Changed:

 - Update iOS/Android/UWP APIs to properly use new API

### Behavioral Changes ###

Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Xamarin.Essentials/Platform/Platform.android.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.android.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.ios.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.ios.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.netstandard.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.netstandard.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.shared.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.shared.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.uwp.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.uwp.cs) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-android.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-android.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-ios.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-ios.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-uwp.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-uwp.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Platform.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/Platform.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C6%5C4%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201806041739022107-277%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=HPELM8LmzhxLxhLEpAPiBDt2NVT7SgTWdUbleZuvtQ0%3D&st=2018-06-04T17%3A35%3A06Z&se=2018-07-05T17%3A40%3A06Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems) | :white_check_mark:Succeeded | |
[DeviceTests/DeviceTests.Shared/Platform_Tests.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/Platform_Tests.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.android.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.android.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.ios.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.ios.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.netstandard.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.netstandard.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.shared.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.shared.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.uwp.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.uwp.cs) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-android.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-android.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-ios.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-ios.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-uwp.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-uwp.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Platform.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/Platform.xml) | :white_check_mark:Succeeded | |
[docs/en/index.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/index.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C6%5C4%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201806041746117994-277%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=9TvCb491ptdeOYKMrVuGFwgZlJ2wqM6RRxf255H5ejs%3D&st=2018-06-04T17%3A42%3A22Z&se=2018-07-05T17%3A47%3A22Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: Xamarin.Essentials/Platform/Platform.android.cs [hunk 1]
```diff
@@ -56,8 +56,25 @@ internal static bool IsIntentSupported(Intent intent)
...
+                return Looper.MyLooper() == Looper.MainLooper;
+            }
+        }
+
         static void PlatformBeginInvokeOnMainThread(Action action)
         {
+            if (IsMainThread)
+            {
+                action();
+                return;
```
Comment (mattleibow) [inline Xamarin.Essentials/Platform/Platform.android.cs]: This is a good change!

File: DeviceTests/DeviceTests.Shared/Platform_Tests.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class Platform_Tests
+    {
+        [Fact]
+        public Task IsOnMainThread()
+        {
+            return Utils.OnMainThread(() =>
+            {
+                Assert.True(Platform.IsMainThread);
+            });
```
Comment (mattleibow) [inline DeviceTests/DeviceTests.Shared/Platform_Tests.cs]: I think we should add a test for background - maybe a `Thread.Run`?

**Comment by mattleibow** [comment]
@jamesmontemagno This looks good, but I think we need a test that runs in the background. Just to make sure we are doing this right.

**Comment by mattleibow** [comment]
Also the `OrientationSensor.xml` docs file is missing - and probably some other docs files.

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems) | :white_check_mark:Succeeded | |
[DeviceTests/DeviceTests.Shared/Platform_Tests.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/Platform_Tests.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.android.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.android.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.ios.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.ios.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.netstandard.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.netstandard.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.shared.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.shared.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.uwp.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.uwp.cs) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-android.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-android.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-ios.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-ios.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-uwp.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-uwp.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensor.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensor.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorChangedEventArgs.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorChangedEventArgs.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorChangedEventHandler.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorChangedEventHandler.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorData.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorData.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Platform.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/Platform.xml) | :white_check_mark:Succeeded | |
[docs/en/index.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/index.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C6%5C4%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201806041828493156-277%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=C9qOru2oif8YfB1Y6FPnl4sBgRyB5NYiCAoLYaUWh2M%3D&st=2018-06-04T18%3A24%3A54Z&se=2018-07-05T18%3A29%3A54Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/DeviceTests.Shared.projitems) | :white_check_mark:Succeeded | |
[DeviceTests/DeviceTests.Shared/Platform_Tests.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/DeviceTests/DeviceTests.Shared/Platform_Tests.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.android.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.android.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.ios.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.ios.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.netstandard.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.netstandard.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.shared.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.shared.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/Platform/Platform.uwp.cs](https://github.com/xamarin/Essentials/blob/feature/mainthread/Xamarin.Essentials/Platform/Platform.uwp.cs) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-android.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-android.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-ios.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-ios.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials-uwp.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials-uwp.xml) | :white_check_mark:Succeeded | |
[docs/en/FrameworksIndex/xamarin-essentials.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/FrameworksIndex/xamarin-essentials.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensor.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensor.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorChangedEventArgs.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorChangedEventArgs.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorChangedEventHandler.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorChangedEventHandler.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/OrientationSensorData.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/OrientationSensorData.xml) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Platform.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/Xamarin.Essentials/Platform.xml) | :white_check_mark:Succeeded | |
[docs/en/index.xml](https://github.com/xamarin/Essentials/blob/feature/mainthread/docs/en/index.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C6%5C4%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201806041831113777-277%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=EOZ4n6jKiKTymOqKx0HQTbLlFYic1Ao%2FowQ%2BUWFPXsM%3D&st=2018-06-04T18%3A27%3A15Z&se=2018-07-05T18%3A32%3A15Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by mattleibow** [comment]
This looks good, but maybe @redth can let us know why the VSTS bot is sad 😢

**Comment by jamesmontemagno** [comment]
So, this one is alright it is the same as Build PR- Win

**Pull request merged**